# Lab_2_Module_3_2. Intro-to-ML

 Challenge 1 - Import and Describe the Dataset


In this challenge we will use the austin_weather.csv file.

First import it into a data frame called austin.

In [100]:
import pandas as pd 
df = pd.read_excel(r'C:\Users\Mtime\OneDrive\Bureau\austin_weather.xlsx')
df.head()

Date  TempHighF  TempAvgF  TempLowF DewPointHighF DewPointAvgF  \
0 2013-12-21         74        60        45            67           49   
1 2013-12-22         56        48        39            43           36   
2 2013-12-23         58        45        32            31           27   
3 2013-12-24         61        46        31            36           28   
4 2013-12-25         58        50        41            44           40   

  DewPointLowF HumidityHighPercent HumidityAvgPercent HumidityLowPercent  ...  \
0           43                  93                 75                 57  ...   
1           28                  93                 68                 43  ...   
2           23                  76                 52                 27  ...   
3           21                  89                 56                 22  ...   
4           36                  86                 71                 56  ...   

  SeaLevelPressureAvgInches SeaLevelPressureLowInches VisibilityHighMiles  \
0                     29.68                     29.59                  10   
1                     30.13                     29.87                  10   
2                     30.49                     30.41                  10   
3                     30.45                      30.3                  10   
4                     30.33                     30.27                  10   

  VisibilityAvgMiles VisibilityLowMiles WindHighMPH WindAvgMPH WindGustMPH  \
0                  7                  2          20          4          31   
1                 10                  5          16          6          25   
2                 10                 10           8          3          12   
3                 10                  7          12          4          20   
4                 10                  7          10          2          16   

  PrecipitationSumInches               Events  
0                   0.46  Rain , Thunderstorm  
1                      0                       
2                      0                       
3                      0                       
4                      T                       

[5 rows x 21 columns]

Next, describe the dataset you have loaded:
    
 . Look at the variables and their types

 . Examine the descriptive statistics of the numeric variables
    
 . Look at the first five rows of all variables to evaluate the categorical variables as well

In [101]:
df.dtypes

Date                          datetime64[ns]
TempHighF                              int64
TempAvgF                               int64
TempLowF                               int64
DewPointHighF                         object
DewPointAvgF                          object
DewPointLowF                          object
HumidityHighPercent                   object
HumidityAvgPercent                    object
HumidityLowPercent                    object
SeaLevelPressureHighInches            object
SeaLevelPressureAvgInches             object
SeaLevelPressureLowInches             object
VisibilityHighMiles                   object
VisibilityAvgMiles                    object
VisibilityLowMiles                    object
WindHighMPH                           object
WindAvgMPH                            object
WindGustMPH                           object
PrecipitationSumInches                object
Events                                object
dtype: object

In [102]:
df.describe()  


TempHighF     TempAvgF     TempLowF
count  1319.000000  1319.000000  1319.000000
mean     80.862775    70.642911    59.902957
std      14.766523    14.045904    14.190648
min      32.000000    29.000000    19.000000
25%      72.000000    62.000000    49.000000
50%      83.000000    73.000000    63.000000
75%      92.000000    83.000000    73.000000
max     107.000000    93.000000    81.000000

Given the information you have learned from examining the dataset, write down three insights about the data in a markdown cell below
Your Insights:

1. There are 21 variables in the dataset. 3 of them are numeric and the rest contain some text.

2. The average temperature in Austin ranged between around 70 degrees F and around 93 degrees F. The highest temperature observed during this period was 107 degrees F and the lowest was 19 degrees F.

3. When we look at the head function, we see that a lot of variables contain numeric data even though these columns are of object type. This means we might have to do some data cleansing.

Let's examine the DewPointAvgF variable by using the unique() function to list all unique values in this dataframe.

Describe what you find in a markdown cell below the code. What did you notice? What do you think made Pandas to treat this column as object instead of int64?



In [103]:
df.DewPointAvgF.unique()

array([49, 36, 27, 28, 40, 39, 41, 26, 42, 22, 48, 32, 8, 11, 45, 55, 61,
       37, 47, 25, 23, 20, 33, 30, 29, 17, 14, 13, 54, 59, 15, 24, 34, 35,
       57, 50, 53, 60, 46, 56, 51, 31, 38, 62, 43, 63, 64, 67, 66, 58, 70,
       68, 65, 69, 71, 72, '-', 73, 74, 21, 44, 52, 12, 75, 76, 18],
      dtype=object)

In [104]:
# in this column we have value like ,'-',  so for this reason pandas take this column like a object type

The following is a list of columns misrepresented as object. Use this list to convert the columns to numeric using the pandas.to_numeric function in the next cell. If you encounter errors in converting strings to numeric values, you need to catch those errors and force the conversion by supplying errors='coerce' as an argument for pandas.to_numeric. Coercing will replace non-convertable elements with NaN which represents an undefined numeric value. This makes it possible for us to conveniently handle missing values in subsequent data processing.

Hint: you may use a loop to change one column at a time but it is more efficient to use apply.

In [105]:
wrong_type_columns = ['DewPointHighF', 'DewPointAvgF', 'DewPointLowF', 'HumidityHighPercent', 
                      'HumidityAvgPercent', 'HumidityLowPercent', 'SeaLevelPressureHighInches', 
                      'SeaLevelPressureAvgInches' ,'SeaLevelPressureLowInches', 'VisibilityHighMiles',
                      'VisibilityAvgMiles', 'VisibilityLowMiles', 'WindHighMPH', 'WindAvgMPH', 
                      'WindGustMPH', 'PrecipitationSumInches']

In [106]:
df[wrong_type_columns] = df[wrong_type_columns].apply(pd.to_numeric, errors='coerce', axis=1)


Check if your code has worked by printing the data types again. You should see only two object columns (Date and Events) now. All other columns should be int64 or float64.

In [107]:
df.dtypes

Date                          datetime64[ns]
TempHighF                              int64
TempAvgF                               int64
TempLowF                               int64
DewPointHighF                        float64
DewPointAvgF                         float64
DewPointLowF                         float64
HumidityHighPercent                  float64
HumidityAvgPercent                   float64
HumidityLowPercent                   float64
SeaLevelPressureHighInches           float64
SeaLevelPressureAvgInches            float64
SeaLevelPressureLowInches            float64
VisibilityHighMiles                  float64
VisibilityAvgMiles                   float64
VisibilityLowMiles                   float64
WindHighMPH                          float64
WindAvgMPH                           float64
WindGustMPH                          float64
PrecipitationSumInches               float64
Events                                object
dtype: object

# Challenge 2 - Handle the Missing Data


By coercing the columns to numeric, we have created NaN for each cell containing characters. We should choose a strategy to address these missing data.

The first step is to examine how many rows contain missing data.

We check how much missing data we have by applying the .isnull() function to our dataset. To find the rows with missing data in any of its cells, we apply .any(axis=1) to the function. austin.isnull().any(axis=1) will return a column containing true if the row contains at least one missing value and false otherwise. Therefore we must subset our dataframe with this column. This will give us all rows with at least one missing value.

In the next cell, identify all rows containing at least one missing value. Assign the dataframes with missing values to a variable called missing_values.

In [108]:
df.isnull().any(axis=1)

0       False
1       False
2       False
3       False
4        True
        ...  
1314    False
1315    False
1316    False
1317    False
1318    False
Length: 1319, dtype: bool

In [109]:
missing_value = df[df.isnull().any(axis=1)]
missing_value

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
4    2013-12-25         58        50        41           44.0          40.0   
6    2013-12-27         60        53        45           41.0          39.0   
7    2013-12-28         62        51        40           43.0          39.0   
42   2014-02-01         76        66        55           62.0          59.0   
51   2014-02-10         60        48        35           49.0          36.0   
...         ...        ...       ...       ...            ...           ...   
1269 2017-06-12         94        85        75           72.0          68.0   
1288 2017-07-01         98        89        79           76.0          72.0   
1304 2017-07-17         98        88        77           75.0          71.0   
1308 2017-07-21        104        91        77           74.0          68.0   
1313 2017-07-26        103        91        78           75.0          70.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
4             36.0                 86.0                71.0   
6             37.0                 83.0                65.0   
7             33.0                 92.0                64.0   
42            41.0                 81.0                71.0   
51            30.0                 82.0                74.0   
...            ...                  ...                 ...   
1269          62.0                 87.0                62.0   
1288          66.0                 85.0                62.0   
1304          66.0                 88.0                63.0   
1308          61.0                 85.0                56.0   
1313          61.0                 91.0                58.0   

      HumidityLowPercent  ...  SeaLevelPressureAvgInches  \
4                   56.0  ...                      30.33   
6                   47.0  ...                      30.39   
7                   36.0  ...                      30.17   
42                  60.0  ...                      29.81   
51                  66.0  ...                      30.15   
...                  ...  ...                        ...   
1269                36.0  ...                      29.89   
1288                38.0  ...                      30.02   
1304                38.0  ...                      29.90   
1308                27.0  ...                      29.95   
1313                25.0  ...                      30.03   

      SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
4                         30.27                 10.0                10.0   
6                         30.34                 10.0                 9.0   
7                         30.04                 10.0                10.0   
42                        29.75                 10.0                10.0   
51                        30.02                 10.0                 8.0   
...                         ...                  ...                 ...   
1269                      29.81                 10.0                10.0   
1288                      29.92                 10.0                 9.0   
1304                      29.84                 10.0                10.0   
1308                      29.86                 10.0                10.0   
1313                      29.93                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
4                    7.0         10.0         2.0         16.0   
6                    7.0          7.0         1.0         11.0   
7                    7.0         10.0         2.0         14.0   
42                   9.0         14.0         6.0         26.0   
51                   4.0         15.0         9.0         23.0   
...                  ...          ...         ...          ...   
1269                 2.0         15.0         7.0         22.0   
1288                 6.0         15.0         7.0         24.0   
1304                10.0         16.0         3.0         23.0   
1308       

There are multiple strategies to handle missing data. Below lists the most common ones data scientists use:

Removing all rows or all columns containing missing data. This is the simplest strategy. It may work in some cases but not others.

Filling all missing values with a placeholder value.

For categorical data, 0, -1, and 9999 are some commonly used placeholder values.
For continuous data, some may opt to fill all missing data with the mean. This strategy is not optimal since it can increase the fit of the model.
Filling the values using some algorithm.

In our case, we will use a hybrid approach which is to first remove the data that contain most missing values then fill in the rest of the missing values with the linear interpolation algorithm.
Next, count the number of rows of austin and missing_values.

In [110]:
df.isnull().sum()

Date                            0
TempHighF                       0
TempAvgF                        0
TempLowF                        0
DewPointHighF                   7
DewPointAvgF                    7
DewPointLowF                    7
HumidityHighPercent             2
HumidityAvgPercent              2
HumidityLowPercent              2
SeaLevelPressureHighInches      3
SeaLevelPressureAvgInches       3
SeaLevelPressureLowInches       3
VisibilityHighMiles            12
VisibilityAvgMiles             12
VisibilityLowMiles             12
WindHighMPH                     2
WindAvgMPH                      2
WindGustMPH                     4
PrecipitationSumInches        124
Events                          0
dtype: int64

In [111]:
df = df.drop('PrecipitationSumInches', axis=1)

In [112]:
df

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0    2013-12-21         74        60        45           67.0          49.0   
1    2013-12-22         56        48        39           43.0          36.0   
2    2013-12-23         58        45        32           31.0          27.0   
3    2013-12-24         61        46        31           36.0          28.0   
4    2013-12-25         58        50        41           44.0          40.0   
...         ...        ...       ...       ...            ...           ...   
1314 2017-07-27        103        89        75           71.0          67.0   
1315 2017-07-28        105        91        76           71.0          64.0   
1316 2017-07-29        107        92        77           72.0          64.0   
1317 2017-07-30        106        93        79           70.0          68.0   
1318 2017-07-31         99        88        77           66.0          61.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
0             43.0                 93.0                75.0   
1             28.0                 93.0                68.0   
2             23.0                 76.0                52.0   
3             21.0                 89.0                56.0   
4             36.0                 86.0                71.0   
...            ...                  ...                 ...   
1314          61.0                 82.0                54.0   
1315          55.0                 87.0                54.0   
1316          55.0                 82.0                51.0   
1317          63.0                 69.0                48.0   
1318          54.0                 64.0                43.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
0                   57.0                       29.86   
1                   43.0                       30.41   
2                   27.0                       30.56   
3                   22.0                       30.56   
4                   56.0                       30.41   
...                  ...                         ...   
1314                25.0                       30.04   
1315                20.0                       29.97   
1316                19.0                       29.91   
1317                27.0                       29.96   
1318                22.0                       30.04   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  \
0                         29.68                      29.59   
1                         30.13                      29.87   
2                         30.49                      30.41   
3                         30.45                      30.30   
4                         30.33                      30.27   
...                         ...                        ...   
1314                      29.97                      29.88   
1315                      29.90                      29.81   
1316                      29.86                      29.79   
1317                      29.91                      29.87   
1318                      29.97                      29.91   

      VisibilityHighMiles  VisibilityAvgMiles  VisibilityLowMiles  \
0                    10.0                 7.0                 2.0   
1                    10.0                10.0                 5.0   
2                    10.0                10.0                10.0   
3                    10.0                10.0                 7.0   
4                    10.0                10.0                 7.0   
...                   ...                 ...                 ...   
1314                 10.0                10.0                10.0   
1315                 10.0                10.0                10.0   
1316                 10.0                10.0                10.0   
1317                 10.0                10.0                10.0   
1318                 10.0                10.0                10.0   

      WindHighMPH  WindAvgMPH  WindGustMPH               E

In [113]:
df = df.drop(['VisibilityHighMiles','VisibilityAvgMiles','VisibilityLowMiles'], axis = 1)

In [114]:
df.isnull().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 7
DewPointAvgF                  7
DewPointLowF                  7
HumidityHighPercent           2
HumidityAvgPercent            2
HumidityLowPercent            2
SeaLevelPressureHighInches    3
SeaLevelPressureAvgInches     3
SeaLevelPressureLowInches     3
WindHighMPH                   2
WindAvgMPH                    2
WindGustMPH                   4
Events                        0
dtype: int64

In [115]:
#df.interpolate(method="polynomial",order=2)
df['DewPointHighF'].interpolate(method='linear', direction = 'forward', inplace=True) 
df['DewPointAvgF'].interpolate(method='linear', direction = 'forward', inplace=True) 
df['DewPointLowF'].interpolate(method='linear', direction = 'forward', inplace=True) 
df['HumidityHighPercent'].interpolate(method='linear', direction = 'forward', inplace=True) 
df['HumidityAvgPercent'].interpolate(method='linear', direction = 'forward', inplace=True) 
df['HumidityLowPercent'].interpolate(method='linear', direction = 'forward', inplace=True) 
df['SeaLevelPressureHighInches'].interpolate(method='linear', direction = 'forward', inplace=True) 
df['WindHighMPH'].interpolate(method='linear', direction = 'forward', inplace=True) 
df['WindAvgMPH'].interpolate(method='linear', direction = 'forward', inplace=True) 
df['WindGustMPH'].interpolate(method='linear', direction = 'forward', inplace=True) 


In [116]:
df['SeaLevelPressureAvgInches'].interpolate(method='linear', direction = 'forward', inplace=True) 
df['SeaLevelPressureLowInches'].interpolate(method='linear', direction = 'forward', inplace=True) 


In [117]:
df.isnull().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
Events                        0
dtype: int64

In [118]:
missing_value

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
4    2013-12-25         58        50        41           44.0          40.0   
6    2013-12-27         60        53        45           41.0          39.0   
7    2013-12-28         62        51        40           43.0          39.0   
42   2014-02-01         76        66        55           62.0          59.0   
51   2014-02-10         60        48        35           49.0          36.0   
...         ...        ...       ...       ...            ...           ...   
1269 2017-06-12         94        85        75           72.0          68.0   
1288 2017-07-01         98        89        79           76.0          72.0   
1304 2017-07-17         98        88        77           75.0          71.0   
1308 2017-07-21        104        91        77           74.0          68.0   
1313 2017-07-26        103        91        78           75.0          70.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
4             36.0                 86.0                71.0   
6             37.0                 83.0                65.0   
7             33.0                 92.0                64.0   
42            41.0                 81.0                71.0   
51            30.0                 82.0                74.0   
...            ...                  ...                 ...   
1269          62.0                 87.0                62.0   
1288          66.0                 85.0                62.0   
1304          66.0                 88.0                63.0   
1308          61.0                 85.0                56.0   
1313          61.0                 91.0                58.0   

      HumidityLowPercent  ...  SeaLevelPressureAvgInches  \
4                   56.0  ...                      30.33   
6                   47.0  ...                      30.39   
7                   36.0  ...                      30.17   
42                  60.0  ...                      29.81   
51                  66.0  ...                      30.15   
...                  ...  ...                        ...   
1269                36.0  ...                      29.89   
1288                38.0  ...                      30.02   
1304                38.0  ...                      29.90   
1308                27.0  ...                      29.95   
1313                25.0  ...                      30.03   

      SeaLevelPressureLowInches  VisibilityHighMiles  VisibilityAvgMiles  \
4                         30.27                 10.0                10.0   
6                         30.34                 10.0                 9.0   
7                         30.04                 10.0                10.0   
42                        29.75                 10.0                10.0   
51                        30.02                 10.0                 8.0   
...                         ...                  ...                 ...   
1269                      29.81                 10.0                10.0   
1288                      29.92                 10.0                 9.0   
1304                      29.84                 10.0                10.0   
1308                      29.86                 10.0                10.0   
1313                      29.93                 10.0                10.0   

      VisibilityLowMiles  WindHighMPH  WindAvgMPH  WindGustMPH  \
4                    7.0         10.0         2.0         16.0   
6                    7.0          7.0         1.0         11.0   
7                    7.0         10.0         2.0         14.0   
42                   9.0         14.0         6.0         26.0   
51                   4.0         15.0         9.0         23.0   
...                  ...          ...         ...          ...   
1269                 2.0         15.0         7.0         22.0   
1288                 6.0         15.0         7.0         24.0   
1304                10.0         16.0         3.0         23.0   
1308       

Calculate the ratio of missing rows to total rows


In [119]:
#flights_num_missing = flights_missing.sum()
len(missing_value)/len(df)

0.10310841546626232

As you can see, there is a large proportion of missing data (over 9%). Perhaps we should evaluate which columns have the most missing data and remove those columns. For the remaining columns, we will perform a linear approximation of the missing data.

We can find the number of missing rows in each column using the .isna() function. We then chain the .sum function to the .isna() function and find the number of missing rows per column

In [120]:
df.isna().sum()

Date                          0
TempHighF                     0
TempAvgF                      0
TempLowF                      0
DewPointHighF                 0
DewPointAvgF                  0
DewPointLowF                  0
HumidityHighPercent           0
HumidityAvgPercent            0
HumidityLowPercent            0
SeaLevelPressureHighInches    0
SeaLevelPressureAvgInches     0
SeaLevelPressureLowInches     0
WindHighMPH                   0
WindAvgMPH                    0
WindGustMPH                   0
Events                        0
dtype: int64

# Challenge 3 - Processing the Events Column


Our dataframe contains one true text column - the Events column. We should evaluate this column to determine how to process it.
Use the value_counts() function to evaluate the contents of this column

In [121]:
df.Events.value_counts()

                             903
Rain                         192
Rain , Thunderstorm          137
Fog , Rain , Thunderstorm     33
Fog                           21
Thunderstorm                  17
Fog , Rain                    14
Rain , Snow                    1
Fog , Thunderstorm             1
Name: Events, dtype: int64

What is the largest number of events happened in a single day? Enter your answer in the next cell.


In [122]:
# the largest event happened ina single day is Rain 

We want to transform the string-type Events values to the numbers. This will allow us to apply machine learning algorithms easily.

How? We will create a new column for each type of events (i.e. Rain, Snow, Fog, Thunderstorm. In each column, we use 1 to indicate if the corresponding event happened in that day and use 0 otherwise.

Below we provide you a list of all event types. Loop the list and create a dummy column with 0 values for each event in austin_fixed. To create a new dummy column with 0 values, simply use austin_fixed[event] = 0.

In [124]:
event_list = ['Snow', 'Fog', 'Rain', 'Thunderstorm']
for item in event_list:
    df[item] = 0
df.head()



Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0 2013-12-21         74        60        45           67.0          49.0   
1 2013-12-22         56        48        39           43.0          36.0   
2 2013-12-23         58        45        32           31.0          27.0   
3 2013-12-24         61        46        31           36.0          28.0   
4 2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  SeaLevelPressureAvgInches  SeaLevelPressureLowInches  WindHighMPH  \
0  ...                      29.68                      29.59         20.0   
1  ...                      30.13                      29.87         16.0   
2  ...                      30.49                      30.41          8.0   
3  ...                      30.45                      30.30         12.0   
4  ...                      30.33                      30.27         10.0   

   WindAvgMPH  WindGustMPH               Events Snow  Fog  Rain  Thunderstorm  
0         4.0         31.0  Rain , Thunderstorm    0    0     0             0  
1         6.0         25.0                         0    0     0             0  
2         3.0         12.0                         0    0     0             0  
3         4.0         20.0                         0    0     0             0  
4         2.0         16.0                         0    0     0             0  

[5 rows x 21 columns]

In [126]:
for item in event_list:
    df[item] = df['Events'].str.contains(item).astype(int)

Print out austin_fixed to check if the event columns are populated with the intended values


In [127]:
df.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0 2013-12-21         74        60        45           67.0          49.0   
1 2013-12-22         56        48        39           43.0          36.0   
2 2013-12-23         58        45        32           31.0          27.0   
3 2013-12-24         61        46        31           36.0          28.0   
4 2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   ...  SeaLevelPressureAvgInches  SeaLevelPressureLowInches  WindHighMPH  \
0  ...                      29.68                      29.59         20.0   
1  ...                      30.13                      29.87         16.0   
2  ...                      30.49                      30.41          8.0   
3  ...                      30.45                      30.30         12.0   
4  ...                      30.33                      30.27         10.0   

   WindAvgMPH  WindGustMPH               Events Snow  Fog  Rain  Thunderstorm  
0         4.0         31.0  Rain , Thunderstorm    0    0     1             1  
1         6.0         25.0                         0    0     0             0  
2         3.0         12.0                         0    0     0             0  
3         4.0         20.0                         0    0     0             0  
4         2.0         16.0                         0    0     0             0  

[5 rows x 21 columns]

If your code worked correctly, now we can drop the Events column as we don't need it any more.


In [129]:
df.drop('Events', axis =1, inplace = True)

In [130]:
df.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0 2013-12-21         74        60        45           67.0          49.0   
1 2013-12-22         56        48        39           43.0          36.0   
2 2013-12-23         58        45        32           31.0          27.0   
3 2013-12-24         61        46        31           36.0          28.0   
4 2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   SeaLevelPressureLowInches  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
0                      29.59         20.0         4.0         31.0     0    0   
1                      29.87         16.0         6.0         25.0     0    0   
2                      30.41          8.0         3.0         12.0     0    0   
3                      30.30         12.0         4.0         20.0     0    0   
4                      30.27         10.0         2.0         16.0     0    0   

   Rain  Thunderstorm  
0     1             1  
1     0             0  
2     0             0  
3     0             0  
4     0             0

# Challenge 4 - Processing The Date Column


In [133]:
from datetime import datetime
df['Date'].apply(lambda x: x.toordinal())

0       735223
1       735224
2       735225
3       735226
4       735227
         ...  
1314    736537
1315    736538
1316    736539
1317    736540
1318    736541
Name: Date, Length: 1319, dtype: int64

In [134]:
df.head()

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
0 2013-12-21         74        60        45           67.0          49.0   
1 2013-12-22         56        48        39           43.0          36.0   
2 2013-12-23         58        45        32           31.0          27.0   
3 2013-12-24         61        46        31           36.0          28.0   
4 2013-12-25         58        50        41           44.0          40.0   

   DewPointLowF  HumidityHighPercent  HumidityAvgPercent  HumidityLowPercent  \
0          43.0                 93.0                75.0                57.0   
1          28.0                 93.0                68.0                43.0   
2          23.0                 76.0                52.0                27.0   
3          21.0                 89.0                56.0                22.0   
4          36.0                 86.0                71.0                56.0   

   SeaLevelPressureHighInches  SeaLevelPressureAvgInches  \
0                       29.86                      29.68   
1                       30.41                      30.13   
2                       30.56                      30.49   
3                       30.56                      30.45   
4                       30.41                      30.33   

   SeaLevelPressureLowInches  WindHighMPH  WindAvgMPH  WindGustMPH  Snow  Fog  \
0                      29.59         20.0         4.0         31.0     0    0   
1                      29.87         16.0         6.0         25.0     0    0   
2                      30.41          8.0         3.0         12.0     0    0   
3                      30.30         12.0         4.0         20.0     0    0   
4                      30.27         10.0         2.0         16.0     0    0   

   Rain  Thunderstorm  
0     1             1  
1     0             0  
2     0             0  
3     0             0  
4     0             0

# Challenge 5 - Sampling and Holdout Sets

Your y should be a subset of austin_fixed containing one column TempAvgF.



In [136]:
subset_df = df[['TempAvgF']]
subset_df.head()

TempAvgF
0        60
1        48
2        45
3        46
4        50

In the next cell, import train_test_split from sklearn.model_selection



In [137]:
from sklearn.model_selection import train_test_split


In [139]:
X_train, X_test, y_train,y_test = train_test_split(df,subset_df, test_size=0.2, random_state=0)

In [140]:
X_train

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
1261 2017-06-04         91        80        68           71.0          69.0   
49   2014-02-08         65        51        37           42.0          36.0   
674  2015-10-26         79        69        58           57.0          54.0   
1191 2017-03-26         86        75        63           66.0          61.0   
549  2015-06-23         92        84        75           76.0          73.0   
...         ...        ...       ...       ...            ...           ...   
763  2016-01-23         57        44        31           32.0          30.0   
835  2016-04-04         83        68        53           53.0          46.0   
1216 2017-04-20         88        79        70           68.0          67.0   
559  2015-07-03         91        84        76           74.0          72.0   
684  2015-11-05         85        79        72           74.0          72.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
1261          66.0                100.0                76.0   
49            29.0                 85.0                63.0   
674           50.0                 78.0                62.0   
1191          51.0                 93.0                67.0   
549           71.0                 94.0                73.0   
...            ...                  ...                 ...   
763           27.0                 85.0                61.0   
835           38.0                 66.0                48.0   
1216          65.0                 90.0                73.0   
559           69.0                 88.0                69.0   
684           69.0                100.0                83.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
1261                52.0                       29.90   
49                  40.0                       30.26   
674                 45.0                       30.05   
1191                40.0                       29.91   
549                 52.0                       30.18   
...                  ...                         ...   
763                 36.0                       30.37   
835                 30.0                       30.21   
1216                55.0                       30.07   
559                 49.0                       30.03   
684                 65.0                       30.02   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  WindHighMPH  \
1261                      29.83                      29.75         12.0   
49                        30.17                      30.03         14.0   
674                       29.97                      29.88         15.0   
1191                      29.84                      29.73         18.0   
549                       30.10                      30.04         13.0   
...                         ...                        ...          ...   
763                       30.28                      30.17          8.0   
835                       30.15                      30.10         12.0   
1216                      30.00                      29.91         12.0   
559                       29.96                      29.89         12.0   
684                       29.97                      29.89         12.0   

      WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
1261         1.0         24.0     0    0     1             1  
49           5.0         21.0     0    0     0             0  
674          6.0         23.0     0    0     0             0  
1191         8.0         27.0     0    0     0             0  
549          5.0         20.0     0    0     0             1  
...          ...          ...   ...  ...   ...           ...  
763          3.0         12.0     0    0     0             0  
835          4.0         18.0     0    0     0             0  
1216         6.0         20.0     0    0     0             0  
559          6.0         19.0     0    0     0             0  
684          6.0         19.0     0   

In [141]:
X_test

Date  TempHighF  TempAvgF  TempLowF  DewPointHighF  DewPointAvgF  \
658  2015-10-10         90        79        67           62.0          59.0   
406  2015-01-31         62        56        49           61.0          51.0   
1202 2017-04-06         79        64        49           47.0          43.0   
202  2014-07-11         95        83        71           72.0          67.0   
1187 2017-03-22         83        76        68           64.0          62.0   
...         ...        ...       ...       ...            ...           ...   
608  2015-08-21         92        82        72           74.0          71.0   
194  2014-07-03         96        84        72           71.0          69.0   
240  2014-08-18        102        90        78           75.0          71.0   
1154 2017-02-17         77        64        50           60.0          50.0   
1275 2017-06-18        100        89        78           74.0          71.0   

      DewPointLowF  HumidityHighPercent  HumidityAvgPercent  \
658           58.0                 78.0                56.0   
406           37.0                100.0                80.0   
1202          40.0                 77.0                52.0   
202           56.0                 90.0                60.0   
1187          58.0                 87.0                67.0   
...            ...                  ...                 ...   
608           67.0                 93.0                73.0   
194           61.0                 93.0                62.0   
240           64.0                 85.0                59.0   
1154          41.0                 93.0                67.0   
1275          67.0                 85.0                60.0   

      HumidityLowPercent  SeaLevelPressureHighInches  \
658                 34.0                       30.21   
406                 59.0                       30.26   
1202                27.0                       30.29   
202                 30.0                       30.15   
1187                47.0                       30.16   
...                  ...                         ...   
608                 52.0                       30.03   
194                 30.0                       30.13   
240                 33.0                       29.99   
1154                41.0                       29.98   
1275                35.0                       29.82   

      SeaLevelPressureAvgInches  SeaLevelPressureLowInches  WindHighMPH  \
658                       30.11                      30.03         10.0   
406                       30.04                      29.90         15.0   
1202                      30.21                      30.16          9.0   
202                       30.09                      30.03         13.0   
1187                      30.07                      30.00         14.0   
...                         ...                        ...          ...   
608                       29.96                      29.86         14.0   
194                       30.08                      30.01         20.0   
240                       29.90                      29.78         15.0   
1154                      29.85                      29.73         13.0   
1275                      29.75                      29.68         13.0   

      WindAvgMPH  WindGustMPH  Snow  Fog  Rain  Thunderstorm  
658          3.0         16.0     0    0     0             0  
406          4.0         21.0     0    0     1             0  
1202         3.0         15.0     0    0     0             0  
202          5.0         22.0     0    0     0             0  
1187         7.0         23.0     0    0     0             0  
...          ...          ...   ...  ...   ...           ...  
608          5.0         23.0     0    0     1             0  
194          3.0         29.0     0    0     1             1  
240          7.0         24.0     0    0     0             0  
1154         6.0         22.0     0    0     0             0  
1275         7.0         23.0     0   